### PR TITLE
Add input_pattern instead of min_pattern_length

### PR DESCRIPTION
**Pull request by zchee**
See also https://github.com/zchee/deoplete-go/issues/7

### CODE REVIEW

**Comment by fatih** [comment]
This didn't work out for me. Now when i put my cursor after a dot, say: `fmt.` it only shows me `Printf(` and nothing else. When I hit `c-x c-o` I see the full list. So it seems `gocode` is not triggered at all.

With deoplete/auto trigger:

image 2016-01-21 at 9 48 33 am

With manual `c-x c-o` keys:

image 2016-01-21 at 9 48 41 am

**Comment by fatih** [comment]
This is the setting I have in my vimrc:

```
let g:deoplete#sources#go = 'vim-go'
let g:deoplete#enable_at_startup = 1
set completeopt+=noinsert
set completeopt+=noselect
```

**Comment by zchee** [comment]
@fatih OK. I will test now.

**Comment by zchee** [comment]
@fatih I testing and debuging now.
It is a tentative movie.

https://asciinema.org/a/34255

``` vim
let g:deoplete#enable_at_startup = 1
let g:deoplete#sources#go = 'vim-go'
set completeopt+=noinsert,noselect
```

It works fine completion after the dot.
Could you try `:UpdateRemotePlugins`?

But, For`fmt` char, There is a case in which occasionally do not completion.
I think caused by plugin manager.

I will continue to debug.

**Comment by svanharmelen** [comment]
@zchee it seems to work perfectly for me :+1:

**Comment by zchee** [comment]
@svanharmelen Thanks for report :)

**Comment by svanharmelen** [comment]
@zchee I think it would be good to set the default rank higher as the default 100. I'm now getting (sometimes) one or two other completion suggestions before the Go completion suggestions.

I know I can change it myself, but it seems a very logical default to me to set it to 1000 by default. Then people still have an option to change it if they do not like the Go completions to show on top.... It seems to make more sense to me.

**Comment by svanharmelen** [comment]
@zchee I also just noticed that it still also suggest completions after typing the `g:deoplete#auto_completion_start_length` amount of characters. That turns out to be pretty annoying as it gives a whole big list of suggestions making it less useful then the old behaviour with just using `deoplete.nvim`.

With `deoplete.nvim` it would **only** call `omni` completions when the pattern was matched. I think this should work the same.

/ping @Shougo for this one as that might be a core `deoplete.nvim` thingy..

**Comment by zchee** [comment]
@svanharmelen Here is `init.vim` and sample code for debugging.
https://gist.github.com/zchee/61aa6347903cb3f41669

> I also just noticed that it still also suggest completions after typing the g:deoplete#auto_completion_start_length amount of characters.

Yes,  I also just noticed.

> That turns out to be pretty annoying as it gives a whole big list of suggestions making it less useful than the old behaviour with just using deoplete.nvim.

I've understood Your explanations. but, Sorry I could not grasp what case :(
Can you tell me what kind of case? or, Could you post sample Go code? :pray:

> With deoplete.nvim it would only call omni completions when the pattern was matched. I think this should work the same.

For example, If use [go-omnifunc-length_0.vim](https://gist.github.com/zchee/61aa6347903cb3f41669#file-go-omnifunc-length_0-vim), it is the same as the state in which the `g:deoplete#auto_completion_start_length` to `0`. 
In this case, `omnifunc` always suggest when the pattern matches.
(I do not yet understand the vim script, So it was difficult to write `1` or more cases.)

This case the same as when using the [go-deoplete-length_0.vim](https://gist.github.com/zchee/61aa6347903cb3f41669#file-go-deoplete-length_0-vim), I think that there is no problem in terms of pattern matches.

I'm sorry if I misunderstood.
WDYT?

**Comment by svanharmelen** [comment]
> I also just noticed that it still also suggest completions after typing the
> g:deoplete#auto_completion_start_length amount of characters.

I think `deoplete-go` completion should **only** be triggered if the pattern matches. It should **not** be triggered after a certain amount of characters.

When using `deoplete.nvim` without `deoplete-go` it also works like that. It only triggers the `omni` source if the pattern maches. So this should work the same I think.

**Comment by svanharmelen** [comment]
If I use these setting it works for me:

```
let g:deoplete#enable_at_startup = 1
let g:deoplete#sources#go = 'vim-go'
call deoplete#custom#set('go', 'rank', 1000)
call deoplete#custom#set('go', 'min_pattern_length', 1000)
```

I made `min_pattern_length` so long it will never trigger. But that feels like a hack. What do you think @Shougo?

**Comment by zchee** [comment]
@svanharmelen Oh, I understand all. Thanks.

The specification of omnifunc has you said method.
and, Certainly I said

>  we not necessary omnifunc.

In https://github.com/zchee/deoplete-go/issues/4#issuecomment-172412821.

But, `deoplete.nvim` and `deoplete-go` is not a substitute of `omnifunc`.
and, It is not intended to realize the same behavior as omnifunc.
The purpose is to call in an asynchronous. and assist in editing the user.

I think whether to the same motion as the `omnifunc` is the thought of users.
I think that it is to solve by writing this note in `README.md`

# 

If you want to the same motion as the `omnifunc`, set

``` vim
call deoplete#custom#set('go', 'min_pattern_length', 1000)
```

# 

WDYT?

**Comment by svanharmelen** [comment]
I guess that would work... It works for me now :wink:

**Comment by zchee** [comment]
@svanharmelen I added notes in README.md.

File: README.md [hunk 1]
```diff
@@ -54,15 +55,47 @@ let g:deoplete#enable_at_startup = 1
...
 
+If you want to the same motion as the `omnifunc`, set
+
+```vim
+call deoplete#custom#set('go', 'min_pattern_length', 1000)
+```
+
+and available setting values,
+
+| value                   | `deoplete.nvim` default                   | `deoplete-go` defalut       |
```
Comment (svanharmelen) [inline README.md]: Small typo: `defalut`

Comment (zchee) [inline README.md]: @svanharmelen :scream_cat:
Fixed.

Comment (svanharmelen) [inline README.md]: :+1:

**Comment by zchee** [comment]
@svanharmelen Merged

**Pull request merged**